### DISCUSSION TITLE
Next js scroll restoration on page refresh

### DISCUSSION THREAD

**Discussion by ivanblank992** [vercel/next.js#26872] [Help]
Hello there,

I am using Next js for building a website. I have read that next js does not support scroll restoration (back to the previous page on the go back button), so I added the experimental feature to the next.config file, and everything seemed to be working. But I want to restore scroll position also on page refresh and that experimental feature did not solve that so I have used a custom hook which I found here on one post.

https://github.com/vercel/next.js/issues/3303
https://github.com/vercel/next.js/pull/22727#issuecomment-803276977
`
import { useEffect } from 'react';

import Router from 'next/router';

const saveScrollPos = (url) => {
  const scrollPos = { x: window.scrollX, y: window.scrollY };
  sessionStorage.setItem(url, JSON.stringify(scrollPos));
};

function restoreScrollPos(url) {
  const scrollPos = JSON.parse(sessionStorage.getItem(url));
  if (scrollPos) {
    window.scrollTo(scrollPos.x, scrollPos.y);
  }
}

const useScrollRestoration = (router) => {
  useEffect(() => {
    if ('scrollRestoration' in window.history) {
      let shouldScrollRestore = false;
      window.history.scrollRestoration = 'manual';
      restoreScrollPos(router.asPath);

      const onBeforeUnload = (event) => {
        saveScrollPos(router.asPath);
        delete event['returnValue'];
      };

      const onRouteChangeStart = () => {
        saveScrollPos(router.asPath);
      };

      const onRouteChangeComplete = (url) => {
        if (shouldScrollRestore) {
          shouldScrollRestore = false;
          restoreScrollPos(url);
        }
      };

      window.addEventListener('beforeunload', onBeforeUnload);
      Router.events.on('routeChangeStart', onRouteChangeStart);
      Router.events.on('routeChangeComplete', onRouteChangeComplete);
      Router.beforePopState(() => {
        shouldScrollRestore = true;
        return true;
      });

      return () => {
        window.removeEventListener('beforeunload', onBeforeUnload);
        Router.events.off('routeChangeStart', onRouteChangeStart);
        Router.events.off('routeChangeComplete', onRouteChangeComplete);
        Router.beforePopState(() => true);
      };
    }
  }, [router]);
};

export default useScrollRestoration;
`

This hook restores scroll on go back button and also maintains scroll position on refresh/reload but it flickers on refresh(you can see in a millisecond top of the page and then the position from where you made a refresh, also when you refresh you do not get same scroll position, it goes a bit down on every refresh(only chrome) so I have added a CSS rule `overflow-anchor: none;` inside Html, body tags and that fixed inconsistent scroll position but not flickering of page top. Does anyone know is there any recommended way to maintain scroll position(restoration) on page refresh and go back button. I do not want to go to the top every time I refresh or go back. Why does native browser scroll restoration works fine on some next js pages without any additional hacks and on some websites like mine not?
Thank you in advance

**Comment by meotimdihia** [comment]
Next.js supported scroll restoration. But if you fetched data on the client and render it. You have to cache the data.
